BUG: Avoid duplicating entire exploded column when joining back with origi…

MarcoGorelli opened the pull request:
…nal frame

- [x] closes #28005
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

jreback commented on pandas/core/frame.py:
```diff
@@ -6240,8 +6240,10 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
         return (
             self.drop([column], axis=1)
-            .join(result)
+            .reset_index(drop=True)  # GH 28005
```
you don’t need the issue number

Commit:
Remove comment with issue number

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6240,8 +6240,10 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
             self.drop([column], axis=1)
-            .join(result)
+            .reset_index(drop=True)
+            .join(self[column].reset_index(drop=True).explode())
```
I don't think we need to `explode` the `Series` twice if you don't drop the index.
```python
result = self[column].reset_index().explode()
return (
    self.drop([column], axis=1)
    .reset_index(drop=True)
    .join(result[column])
    .reindex(columns=self.columns, copy=False)
    .set_index(result[index_col])
)
```

I guess it's not immediate to get `index_col` since the index can have a name or not, but I think it's worth avoiding the double `explode`.

MarcoGorelli replied:
Thanks for your review!

That's a good point, thanks - I'll work on this

datapythonista reviewed (changes requested):
Looks good, just one comment for performance.

Commit:
Rewrite
Syntax: python explode.py infile template [range]

The template argument is used to construct the names of the
individual frame files.  The frames are numbered file001.ext,
file002.ext, etc.  You can insert %d to control the placement
and syntax of the frame number.

The optional range argument specifies which frames to extract.
You can give one or more ranges like 1-10, 5, -15 etc.  If
omitted, all frames are extracted. method of frame so it doesn't require exploding a series twice

Commit:
Remove commas after kwargs from autoformatting

Commit:
Remove more trailing commas

Commit:
Rewrite how string is formatted (for backwards compatibility)

Commit:
Add multiindex (named and unnamed) to test cases

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6237,13 +6237,31 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+        if isinstance(self.index, MultiIndex):
+            index_names = [
+                "level_{num}".format(num=num) if val is None else val
+                for num, val in enumerate(self.index.names)
```
I think you don't care about the names if you set the index after the join with something like:
```python
result.index = pandas.MultiIndex.from_frame(result.iloc[:, :self.index.nlevels])
```

I think this will make this code a bit simpler.

MarcoGorelli replied:
I may have misunderstood, but are you sure we can disregard names? If the original index is unnamed, then after `reset_index()`, it will be named `index`. 

E.g.:
```
import pandas as pd
df = pd.DataFrame([1,2,3]).reset_index()
df.iloc[:, :df.index.nlevels]
```
>   index
>0      0
>1      1
>2      2

datapythonista replied:
You're setting the index at the very end to the original one, you'll have to leave that.

MarcoGorelli replied:
I've pushed a simpler solution which uses the change you suggested - thanks for taking the time to review my work!

datapythonista reviewed (commented):
couple of comments

MarcoGorelli commented on pandas/core/frame.py:
```diff
@@ -6237,13 +6237,31 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+                for num, val in enumerate(self.index.names)
+            ]
+        else:
+            index_names = [i if i else "index" for i in [self.index.name]]
```
Thanks, this is much cleaner!

pep8speaks commented:
Hello @MarcoGorelli! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pandas/core/frame.py`](https://github.com/pandas-dev/pandas/blob/44fb0640613aac6083a220474f171f2d08f88ea3/pandas/core/frame.py):

> [Line 6251:89](https://github.com/pandas-dev/pandas/blob/44fb0640613aac6083a220474f171f2d08f88ea3/pandas/core/frame.py#L6251): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)
> [Line 6256:40](https://github.com/pandas-dev/pandas/blob/44fb0640613aac6083a220474f171f2d08f88ea3/pandas/core/frame.py#L6256): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator

Commit:
Avoid having to keep track of index names

Commit:
Black autoformatting

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6237,13 +6237,30 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+            result.index = result.iloc[:, 0]
+
+        if not named_index:
+            result.index.names = [None] * self.index.nlevels
```
Didn't run this, but I have the feeling that if the names of the MultiIndex levels in the input DataFrame are something like `['foo', None, 'bar']`, in the output will be `['foo', 'level_0', 'bar']`. If that's the case, besides fixing it, would be nice to add an example.

jreback replied:
we don’t implicitly ever change names like this

MarcoGorelli replied:
@datapythonista you're right. I didn't think partially named multiindices were a thing, sorry for not having checked. Fix and test cases added.

@jreback sorry about that - should it be `result.index.rename()` instead?

Thanks to both of you for your time.

By the way, even on the master branch (without any of my changes applied), this sometimes fails:
```
pytest pandas/tests/computation/test_eval.py -k test_binary_arith_ops
```
I just ran it 10 times, and 6 of them had a failure.

From what I can discern from the Travis CI build logs, that's what's currently failing here

Commit:
Add fix and test for case of partially named multiindex

Commit:
Black autoformatting

Commit:
Clean up

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6243,13 +6243,28 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+                result.iloc[:, : self.index.nlevels]
+            )
+            result.index.rename(
+                [i or None for num, i in enumerate(self.index.names)], inplace=True
```
I'm sure I'm missing something, but why can't this be simply `result.index.names = self.index.names`?

MarcoGorelli replied:
No reason :) Change pushed, thanks for the suggestion!

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6243,13 +6243,28 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+            )
+        else:
+            result.index = result.iloc[:, 0]
+            result.index.rename(self.index.name or None, inplace=True)
```
same

Commit:
Simplify renaming index

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6243,13 +6243,24 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
-            .join(result)
-            .reindex(columns=self.columns, copy=False)
+            .reset_index()
+            .join(column_with_index[column].explode())
```
I'm thinking that all this is the same as:
```python
exploded_col = self.pop(column).reset_index(drop=True).explode()
self.reset_index().join(exploded_col)
```

We probably don't want to modify self but a copy, because we need `self.columns` later, but I think that is a bit clearer.

datapythonista reviewed (commented):
looks good, just one final suggestion

Commit:
Re-write in more readable way

Commit:
Remove old whatsnew entry

Commit:
Use ABCMultiindex to assert type of index

Commit:
Add whatsnew entry

TomAugspurger commented on doc/source/whatsnew/v0.25.2.rst:
```diff
@@ -49,7 +49,7 @@ Interval
 ^^^^^^^^
 
--
+- Frame gets duplicated when exploding a single column and there are duplicates in the index (:issue:`28010`).
```
This should go in v1.0.0.rst.

Commit:
Add whatsnew entry in v1.0.0.rst

datapythonista reviewed (commented):
Just two last small things, and from my side looks good. Thanks for the patience on this.

Commit:
Rephrase whatsnew entry, name copy of self result rather than frame

datapythonista reviewed (approved):
lgtm, thanks @MarcoGorelli

MarcoGorelli commented:
Thanks for your help Marc, really appreciate it!

jreback commented on pandas/core/frame.py:
```diff
@@ -6244,12 +6244,20 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+        exploded_col = result.pop(column).reset_index(drop=True).explode()
+        result = result.reset_index().join(exploded_col)
+
+        if isinstance(self.index, ABCMultiIndex):
```
this is not great here, what exactly are you trying to do?

MarcoGorelli replied:
I'm trying to check whether the index is a multiindex or not - saw this in other parts of the module, e.g.
```
                if isinstance(self.index, ABCMultiIndex):
                    # array of tuples to numpy cols. copy copy copy
                    ix_vals = list(map(np.array, zip(*self.index.values)))
                else:
                    ix_vals = [self.index.values]
```

jreback replied:
of course, that’s not what i mean

it is completely o scrub what you are doing
pls show me an example of the index before and after for a MI and an Indez

the purpose of the .explode()  is to exactly handle this
this likely needs to be directly handled in the cython code instead

jreback replied:
we do not want to special case MI (other codes are generally just older )

datapythonista replied:
not sure I understand what you're proposing @jreback. The original implementation calls `explode` in the `Series` as it does in here, so it's calling the cython version, but when joining it joins by the index and if there are duplicate entries in the index it does an unwanted cross-product.

Actually, it just came to my mind, that we could make things much simpler if instead of resetting the index to join on it without duplicates, we create a temporary column with a range and we join with it. That would make things much simpler and less hacky, and there is no need to touch the index. Does that make sense to you?

MarcoGorelli replied:
@jreback thanks for your review - I've removed the special case handling

@datapythonista where would that temporary column go though (and how would I name it without risking overwriting an existing column)? I can't add this temporary range column to `self` _before_ calling `explode`, and then I would be exploding a dataframe, and that's exactly the method I'm trying to fix, and I don't see how adding it after helps

datapythonista replied:
you're right, what I said doesn't make sense, I got confused, sorry

Commit:
Remove special-casing multiindex

jreback commented on pandas/core/frame.py:
```diff
@@ -6244,12 +6244,14 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
+        result = self.copy()
+        exploded_col = result.pop(column).reset_index(drop=True).explode()
+        result = result.reset_index().join(exploded_col)
+        result.set_index(result.columns[: self.index.nlevels].tolist(), inplace=True)
```
you don't need to do all of this, try
```result.index = self.index.take(result.index)```

MarcoGorelli replied:
Nice, I didn't know about `take`, thanks Jeff!

jreback commented on pandas/core/frame.py:
```diff
@@ -6244,12 +6244,14 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
-            .join(result)
-            .reindex(columns=self.columns, copy=False)
-        )
+        result = self.copy()
```
you can start this off by doing
```
df = self.reset_index(drop=True)
result = df[column].explode()
result = (df.drop([column], axis=1)
       .join(result)
result.index = self.index.take(result.index)
....
```

MarcoGorelli committed:
further simplification using take

TomAugspurger commented:
@MarcoGorelli looks like a git merge issue. Can you try fetching and merging master, and then repushing? Let us know if you get stuck.

datapythonista commented on pandas/core/frame.py:
```diff
@@ -6259,12 +6259,13 @@ def explode(self, column: Union[str, Tuple]) -> "DataFrame":
-        )
+        df = self.reset_index(drop=True)
+        result = df[column].explode()
+        result = df.drop([column], axis=1).join(result)
```
I think the naming here is a bit confusing. `result` is first a `Series`, and then it's this `DataFrame`. I don't think the previous `result` is being named correctly.

Also thinking if this can be implemented with method chaining, I think that usually improves readability. I mean something like:
```python
df =self.reset_index(drop=True)
return (
    df.drop([column], axis=1)
      .join(df[column].explode())
      .set_index(self.index.take(df.index))
      .reindex(columns=self.columns, copy=False))
```

In any case, this is looking much better at every iteration.

MarcoGorelli replied:
Yes, that's better - I still need to break it up slightly though, don't I? Else in `self.index.take(df.index)`, `df` isn't the right length.

Anyway, naming cleared up

MarcoGorelli committed:
Fix merge conflict in whatsnew

MarcoGorelli committed:
Remove repetitive commands in test parametrization

MarcoGorelli committed:
Further reformatting

MarcoGorelli commented:
@TomAugspurger sorry to chase you up on this - I've resolved the merge conflict, is there anything else you'd like changed?

TomAugspurger commented:
Thanks!
